crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+
+/// Classification for the casing of an element in a
+/// sequence of literal strings.
+///
+/// This is necessary to achieve an "isort-style" sort,
+/// where elements are sorted first by category,
+/// then, within categories, are sorted according
+/// to a natural sort.
+///
+/// You'll notice that a very similar enum exists
+/// in ruff's reimplementation of isort.
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone, Copy)]
+enum InferredMemberType {
+    Constant,
+    Class,
+    Other,
+}
+
+impl InferredMemberType {
+    fn of(value: &str) -> Self {
+        // E.g. `CONSTANT`
+        if value.len() > 1 && is_cased_uppercase(value) {
+            Self::Constant
+        // E.g. `Class`
+        } else if value.starts_with(char::is_uppercase) {
+            Self::Class
+        // E.g. `some_variable` or `some_function`
+        } else {
+            Self::Other
+        }
+    }
+}
+
+/// An enumeration of the various kinds of sequences for which Python has
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries).
+///
+/// (I'm aware a set isn't actually a "sequence",
+/// *but* for our purposes it's conceptually a sequence,
+/// since in terms of the AST structure it's almost identical
+/// to tuples/lists.)
+///
+/// Whereas lists, dicts and sets are always parenthesized
+/// (e.g. lists always start with `[` and end with `]`),
+/// single-line tuples *can* be unparenthesized.
+/// We keep the original AST node around for the
+/// Tuple variant so that this can be queried later.
+#[derive(Debug)]
+pub(super) enum SequenceKind<'a> {
+    List,
+    Set,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl SequenceKind<'_> {
+    fn surrounding_parens(&self, source: &str) -> (&'static str, &'static str) {
+        match self {
+            Self::List => ("[", "]"),
+            Self::Set => ("{", "}"),
+            Self::Tuple(ast_node) => {
+                if ast_node.is_parenthesized(source) {
+                    ("(", ")")
+                } else {
+                    ("", "")
+                }
+            }
+        }
+    }
+
+    const fn opening_token_for_multiline_definition(&self) -> TokenKind {
+        match self {
+            Self::List => TokenKind::Lsqb,
+            Self::Set => TokenKind::Lbrace,
+            Self::Tuple(_) => TokenKind::Lpar,
+        }
+    }
+
+    const fn closing_token_for_multiline_definition(&self) -> TokenKind {
+        match self {
+            Self::List => TokenKind::Rsqb,
+            Self::Set => TokenKind::Rbrace,
+            Self::Tuple(_) => TokenKind::Rpar,
+        }
+    }
+}
+
+/// An enumeration of the various kinds of
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries)
+/// Python provides for builtin containers.
+#[derive(Debug, is_macro::Is)]
+pub(super) enum DisplayKind<'a> {
+    Sequence(SequenceKind<'a>),
+    Dict { values: &'a Vec<ast::Expr> },
```

Comment:
You typically want `&[ast::Expr]` instead of `&Vec<ast::Expr>`. There are other ways to create a slice, and any function that accepts `&[Expr]` will accept `&Vec<Expr>`, but if you have an `&[Expr]`, you can't convert it the other way to `&Vec<Expr>` without allocating. So, you likely want the most permissive type here (`&[Expr]`). It's similar to `&String` vs. `&str`.